public/app/plugins/panel/heatmap-new/utils.ts

```diff
@@ -460,3 +470,64 @@ export const countsToFills = (u: uPlot, seriesIdx: number, palette: string[]) =>
 
   return indexedFills;
 };
+
+export const getHeatmapFields = (dataFrame: DataFrame): Array<Field | undefined> => {
+  const xField: Field | undefined = dataFrame.fields.find((f) => f.name === 'xMin');
+  const yField: Field | undefined = dataFrame.fields.find((f) => f.name === 'yMin');
+  const countField: Field | undefined = dataFrame.fields.find((f) => f.name === 'count');
+
+  return [xField, yField, countField];
+};
+
+export const getHeatmapArrays = (dataFrame: DataFrame): Array<number[] | undefined> => {
+  const [xField, yField, countField] = getHeatmapFields(dataFrame);
+  return [xField?.values.toArray(), yField?.values.toArray(), countField?.values.toArray()];
+};
+
+export const getDataMapping = (
+  heatmapData: HeatmapData,
```

Comment:
@srclosson -- do we only ever use `heatmapData.heatmap`?  If so, lets make the DataFrame the only input variable

Reply:
No, we use the geometry to get max values.

Reply:
```  
  const xsmax = fxs.values.get(fxs.values.length - 1) + heatmapData.xBucketSize!;
  const ysmax = fys.values.get(fys.values.length - 1) + heatmapData.yBucketSize!;
```

as well as 
```
    const xBucketIdx = Math.floor(incrRoundDn(incrRoundUp((xo - xsmin) / heatmapData.xBucketSize!, 1e-7), 1e-7));
    const yBucketIdx = Math.floor(incrRoundDn(incrRoundUp((yo - ysmin) / heatmapData.yBucketSize!, 1e-7), 1e-7));
```

Reply:
Although, these values could be derived...